File: WoofAndroid/presentation/profile/src/main/kotlin/com/perrystreet/woof/presentation/profile/viewmodel/ProfileModerationViewModel.kt
package com.perrystreet.woof.presentation.profile.viewmodel

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.presentation.common.viewmodel.StateProducingViewModel
import com.perrystreet.woof.presentation.navigation.INavigator
import com.perrystreet.woof.presentation.profile.uimodel.ProfileModerationDialogUIModel
import com.perrystreet.woof.presentation.profile.uimodel.ProfileToastUIModel
import com.perrystreet.woof.usecase.moderation.BlockDogUseCase
import com.perrystreet.woof.usecase.moderation.ReportDogUseCase
import com.perrystreet.woof.utils.guard
import io.reactivex.rxjava3.kotlin.plusAssign
import org.koin.android.annotation.KoinViewModel
import org.koin.core.annotation.InjectedParam
import java.util.Optional

@KoinViewModel
class ProfileModerationViewModel(
    @InjectedParam private val dog: Dog,
    private val reportDogUseCase: ReportDogUseCase,
    private val blockDogUseCase: BlockDogUseCase,
    private val navigator: INavigator,
) : StateProducingViewModel<ProfileModerationViewModel.State>(initialValue = State.Initial) {

    data class State(
        val dialog: ProfileModerationDialogUIModel?,
        val toast: ProfileToastUIModel?,
    ) {
        companion object {
            val Initial = State(dialog = null, toast = null)
        }
    }

    fun onReportTap() {
        _state.onNext(currentState.copy(dialog = ProfileModerationDialogUIModel.Report(dog.name)))
    }

    fun onBlockTap() {
        _state.onNext(currentState.copy(dialog = ProfileModerationDialogUIModel.Block(dog.name)))
    }

    fun onDialogDismiss() {
        _state.onNext(currentState.copy(dialog = null))
    }

    fun onDialogConfirm() {
        val dialog = currentState.dialog
        guard(dialog != null) { return }
        _state.onNext(currentState.copy(dialog = null))
        when (dialog) {
            is ProfileModerationDialogUIModel.Report -> report()
            is ProfileModerationDialogUIModel.Block -> block()
        }
    }

    fun onToastDismiss() {
        _state.onNext(currentState.copy(toast = null))
    }

    private fun report() {
        disposables += reportDogUseCase(dog).subscribe(
            { _state.onNext(currentState.copy(toast = ProfileToastUIModel.ReportSent)) },
            { error -> mutableError.onNext(Optional.of(error)) },
        )
    }

    private fun block() {
        disposables += blockDogUseCase(dog).subscribe(
            { navigator.back() },
            { error -> mutableError.onNext(Optional.of(error)) },
        )
    }
}

File: WoofIOS/SwiftPackages/Presentation/Profile/Sources/PresentationProfile/ViewModel/ProfileModerationViewModel.swift
import Combine
import DI
import Models
import PresentationCommon
import PresentationNavigation
import UseCase
import Utils

@Factory
public final class ProfileModerationViewModel: StateProducingViewModel<ProfileModerationViewModel.State, ModerationError> {
    @DIArgument private let dog: Dog
    private let reportDogUseCase: ReportDogUseCase
    private let blockDogUseCase: BlockDogUseCase
    private let navigator: NavigatorImplementing

    init(
        dog: Dog,
        reportDogUseCase: ReportDogUseCase,
        blockDogUseCase: BlockDogUseCase,
        navigator: NavigatorImplementing
    ) {
        self.dog = dog
        self.reportDogUseCase = reportDogUseCase
        self.blockDogUseCase = blockDogUseCase
        self.navigator = navigator
        super.init(initialValue: .initial)
    }

    public func onReportTap() {
        setState(currentState.copy(dialog: .report(name: dog.name)))
    }

    public func onBlockTap() {
        setState(currentState.copy(dialog: .block(name: dog.name)))
    }

    public func onDialogDismiss() {
        setState(currentState.copy(dialog: .some(nil)))
    }

    public func onDialogConfirm() {
        guard let dialog = currentState.dialog else { return }
        setState(currentState.copy(dialog: .some(nil)))
        switch dialog {
        case .report: report()
        case .block: block()
        }
    }

    public func onToastDismiss() {
        setState(currentState.copy(toast: .some(nil)))
    }

    private func report() {
        reportDogUseCase(dog: dog)
            .pss_sink { [weak self] result in
                guard let self else { return }
                switch result {
                case .success: setState(currentState.copy(toast: .reportSent))
                case .failure(let error): emitError(error)
                }
            }
            .store(in: &cancellables)
    }

    private func block() {
        blockDogUseCase(dog: dog)
            .pss_sink { [weak self] result in
                guard let self else { return }
                switch result {
                case .success: navigator.back()
                case .failure(let error): emitError(error)
                }
            }
            .store(in: &cancellables)
    }
}

extension ProfileModerationViewModel {
    public struct State: Equatable {
        public let dialog: ProfileModerationDialogUIModel?
        public let toast: ProfileToastUIModel?

        public static let initial = State(dialog: nil, toast: nil)

        public init(dialog: ProfileModerationDialogUIModel?, toast: ProfileToastUIModel?) {
            self.dialog = dialog
            self.toast = toast
        }

        func copy(dialog: ProfileModerationDialogUIModel?? = nil, toast: ProfileToastUIModel?? = nil) -> State {
            State(dialog: dialog ?? self.dialog, toast: toast ?? self.toast)
        }
    }
}
